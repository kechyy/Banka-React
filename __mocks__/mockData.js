import jwt from 'jsonwebtoken';

const secret = 'nkechi';

export const userToken = `Bearer ${jwt.sign(
  {
    id: 1,
    email: 'example@gmail.com',
    type: 'user'
  },
  secret,
  { expiresIn: '1 hour' }
)}`;

export const expiredToken = `Bearer ${jwt.sign(
  {
    id: 1,
    email: 'example@gmail.com',
    type: 'user'
  },
  secret,
  { expiresIn: -1 }
)}`;

export default {
  authResponse: {
    data: {
      id: 1,
      username: 'Kech123',
      email: 'kech123@andela.com',
      type: 'user',
      token: userToken
    }
  }
};