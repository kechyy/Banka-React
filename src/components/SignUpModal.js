import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { modalToggle, modalToggle2, closeModal} from '../store/modules/modals/actions';
import { authAction } from '../store/modules/auth/actions';
import checkUserType from '../utils/checkUserType'

class SignUpModal extends Component {
  constructor() {
    super();
    this.handleInputFieldChange = this.handleInputFieldChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.isLoading = false;
    this.state = {
      submittting: false,
      firstName: '',
      lastName: '',
      email: '',
      password: '',
      cPassword: '',
      errors: {
        firstNmae: '',
        lastName: '',
        email: '',
        password: '',
        cpassword: ''
      }
    };
  }
  handleSubmit(e) {
    e.preventDefault();
    const { firstName, lastName, email, password, cpassword } = this.state;

    const { authAction: signinAction } = this.props;
    const errors = {};
    if (!firstName) errors.firstName = 'Firstname cannot be empty';
    if (!lastName) errors.lastName = 'Lastname cannot be empty';
    if (!email) errors.email = 'Email cannot be empty';
    if (!password) errors.password = 'Password cannot be empty';
    if (!cpassword) errors. cpassword = 'Comfirm password cannot be empty';
    if (Object.keys(errors).length) {
      this.setState(prevState => ({
        ...prevState,
        errors: {
          ...errors
        }
      }));
      return;
    }
    const {
      history
    } = this.props;
    const  url = checkUserType();
    // this.setState({ submittting: true });
    signinAction({ userData: { firstName, lastName, email, password, cpassword }, history, url });
  }

  handleInputFieldChange(event) {
    const {
      target: { name, value }
    } = event;
    this.setState(prevState => ({
      [name]: value,
      errors: {
        ...prevState.errors,
        [name]: ''
      }
    }));
  }

  render() {
    const {
      errors: { firstName: firstNameError, lastName: lastNameError,
        email: emailError, password: passwordError , cpassword: cpasswordError }
    } = this.state;
    const {closeModal, signupDisplay, page, modalToggle, modalToggle2 } = this.props;
    
    return (
      <div className="modal" style={{display: signupDisplay}}>
        <div className="modal-content">
          <header>
            <a href="javascript:;" title="Close" className="modal-close" onClick={closeModal}>&times;</a>
            <h1>USER SIGN UP PAGE</h1>
            <p>Enter Your Sign-Up Details To Create An Account</p>
          </header>
          <form onSubmit={this.handleSubmit}>
          <div className="modal-body">
            <div className="row loginForm">
              <label>First Name</label>
              <div className="col-xs-12 col-sm-12 col-lg-12">
                <input type="text" className="input-block" name="firstName"
                 id="firstName" placeholder="Enter first name" required onChange={this.handleInputFieldChange} />
                {!!firstNameError && (
                  <span className="text-red-600 text-xs">{firstNameError}</span>
                )}
              </div>
              <label>Last Name</label>
              <div className="col-xs-12 col-sm-12 col-lg-12">
                <input type="text" className="input-block"  name="lastName"
                id="lastName" placeholder="Enter Last name" required onChange={this.handleInputFieldChange} />
                {!!lastNameError && (
                  <span className="text-red-600 text-xs">{lastNameError}</span>
                )}
              </div>
              <label>Email Address</label>
              <div className="col-xs-12 col-sm-12">
                <input type="email" className="input-block" name="email" 
                 id="email" placeholder="Enter email address" required onChange={this.handleInputFieldChange} />
                {!!emailError && (
                  <span className="text-red-600 text-xs">{emailError}</span>
                )}
              </div>
              <label>Password</label>
              <div className="col-xs-12 col-sm-12 col-lg-12">
                <input type="password" className="input-block" required name="password"
                 id="password" placeholder="Enter password" onChange={this.handleInputFieldChange} />
                 {!!passwordError && (
                  <span className="text-red-600 text-xs">{passwordError}</span>
                )}
                </div>
                <label>Confirm Password</label>
                <div className="col-sx-12 col-sm-12 col-lg-12">
                  <input type="password" className="input-block" required name="cpassword"
                  id="cpassword" placeholder="Enter password" onChange={this.handleInputFieldChange} />
                  {!!cpasswordError && (
                    <span className="text-red-600 text-xs">{cpasswordError}</span>
                  )}
                </div>
                <div className="col-xs-12 col-sm-12 col-lg-12">
                <button className="btn btn-pink btn-full"  type="submit" >Sign Up</button><br/>
                <br/>
                <a href="javascript:;" className="sign" onClick={modalToggle}>Don't have an account? <span  className="btn-yellow-test">[SIGN IN]</span></a>
              </div>
            </div>
          </div> 
          </form> 
        </div>
      </div>
    )
  }
}
  
const mapStateToProps = state => ({
  signupDisplay: state.modalStatus.signupDisplay,
  page: state.modalStatus.page
});
const mapDispatchToProps = {
  modalToggle, modalToggle2, closeModal, authAction
}
export default withRouter(connect(mapStateToProps, mapDispatchToProps)(SignUpModal));