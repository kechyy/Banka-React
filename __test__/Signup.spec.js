import React from 'react';
import { mount, shallow } from './enzyme';
import toJson from 'enzyme-to-json';
import { Provider } from 'react-redux';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import SignUpModal from '../src/components/SignUpModal';

import { userToken } from '../__mocks__/mockData';

const mockStore = configureMockStore([thunk]);
const renderWithEnzymes = state => {
  const store = mockStore({
    authReducer: {
      error: null,
      isAuthenticated: false,
      user: {},
      status: 'rest'
    },
    modalStatus: {
      signinDisplay: 'none',
      signupDisplay: 'none',
      page: '',
    },
    SignUpModal: {
      ...state
    }
  });
  return mount(
    <Provider store={store}>
      <Router>
        <SignUpModal {...props} />
      </Router>
    </Provider>
  );
};

const props = {
  authAction: jest.fn(),
  status: 'rest',
  isAuthenticated: false,
  error: null,
  location: { url: '/user-dashboard' },
  history: { push: jest.fn() },
  closeModal:jest.fn(),
  modalToggle2:jest.fn(),
  modalToggle:jest.fn(),
  signupDisplay:'block',
  page:'signup',
  signinDisplay: 'none'

};


describe('Signup component', () => {
 
  it('renders without crashing', () => {
   const wrapper =renderWithEnzymes();

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders one form', () => {
    const wrapper =renderWithEnzymes();




    expect(wrapper.find('h1').text()).toEqual('USER SIGN UP PAGE');
    expect(wrapper.find('SignUpModal')).toBeTruthy();
    expect(wrapper.find('form')).toBeTruthy();
    expect(wrapper.find('form').length).toBe(1);
  });

  it('renders 4 input fields and 1 submit button ', () => {
    const wrapper =renderWithEnzymes();

    // console.log('new wrapper >>', wrapper.debug());

    expect(wrapper.find('input').length).toBe(5);
    expect(wrapper.find('input').at(0).props().type).toBe('text');
    expect(wrapper.find('input').at(1).props().type).toBe('text');
    expect(wrapper.find('input').at(2).props().type).toBe('email');
    expect(wrapper.find('input').at(3).props().type).toBe('password');
    expect(wrapper.find('input').at(4).props().type).toBe('password');
    expect(wrapper.find('button').length).toBe(1);
    expect(wrapper.find('button').at(0).props().type).toBe('submit');
  });

  it('should submit a valid form', () => {
    const wrapper = renderWithEnzymes();
    const event = {
      preventDefault: jest.fn()
    };
    const typeInputFirstName = wrapper.find("input[name='firstName']");
    typeInputFirstName.simulate('change', {
      target: {
        name: 'firstName',
        value: 'Nkechi'
      }
    });
    const typeInputlastName = wrapper.find("input[name='lastName']");
    typeInputlastName.simulate('change', {
      target: {
        name: 'lastName',
        value: 'Ogbonna'
      }
    });
    const typeInputEmail = wrapper.find("input[name='email']");
    typeInputEmail.simulate('change', {
      target: {
        name: 'email',
        value: 'kech@gmail.com'
      }
    });
    const typeInputPassword = wrapper.find("input[name='password']");
    typeInputPassword.simulate('change', {
      target: {
        name: 'password',
        value: 'kech@123'
      }
    });
    const typeInputCPassword = wrapper.find("input[name='cpassword']");
    typeInputCPassword.simulate('change', {
      target: {
        name: 'cpassword',
        value: 'kech@123'
      }
    });
    
    wrapper.find('form').simulate('submit', event);
    expect(wrapper.find('SignUpModal').state('firstName')).toEqual('Nkechi');
    expect(wrapper.find('SignUpModal').state('lastName')).toEqual('Ogbonna');
    expect(wrapper.find('SignUpModal').state('email')).toEqual('kech@gmail.com');
    expect(wrapper.find('SignUpModal').state('password')).toEqual('kech@123');
    expect(wrapper.find('SignUpModal').state('cpassword')).toEqual('kech@123');
  });
})