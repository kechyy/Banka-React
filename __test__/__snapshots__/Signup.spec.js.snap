// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup component renders without crashing 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <withRouter(Connect(SignUpModal))
        authAction={[MockFunction]}
        closeModal={[MockFunction]}
        error={null}
        history={
          Object {
            "push": [MockFunction],
          }
        }
        isAuthenticated={false}
        location={
          Object {
            "url": "/user-dashboard",
          }
        }
        modalToggle={[MockFunction]}
        modalToggle2={[MockFunction]}
        page="signup"
        signinDisplay="none"
        signupDisplay="block"
        status="rest"
      >
        <Connect(SignUpModal)
          authAction={[MockFunction]}
          closeModal={[MockFunction]}
          error={null}
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "createHref": [Function],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          isAuthenticated={false}
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
          modalToggle={[MockFunction]}
          modalToggle2={[MockFunction]}
          page="signup"
          signinDisplay="none"
          signupDisplay="block"
          status="rest"
        >
          <SignUpModal
            authAction={[Function]}
            closeModal={[Function]}
            error={null}
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "createHref": [Function],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            isAuthenticated={false}
            location={
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              }
            }
            match={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
            modalToggle={[Function]}
            modalToggle2={[Function]}
            page=""
            signinDisplay="none"
            signupDisplay="none"
            status="rest"
          >
            <div
              className="modal"
            >
              <div
                className="modal-content"
              >
                <header>
                  <h1>
                    USER SIGN UP PAGE
                  </h1>
                  <p>
                    Enter Your Sign-Up Details To Create An Account
                  </p>
                </header>
                <form
                  onSubmit={[Function]}
                >
                  <div
                    className="modal-body"
                  >
                    <div
                      className="row loginForm"
                    >
                      <label>
                        First Name
                      </label>
                      <div
                        className="col-xs-12 col-sm-12 col-lg-12"
                      >
                        <input
                          className="input-block"
                          id="firstName"
                          name="firstName"
                          onChange={[Function]}
                          placeholder="Enter first name"
                          required={true}
                          type="text"
                        />
                      </div>
                      <label>
                        Last Name
                      </label>
                      <div
                        className="col-xs-12 col-sm-12 col-lg-12"
                      >
                        <input
                          className="input-block"
                          id="lastName"
                          name="lastName"
                          onChange={[Function]}
                          placeholder="Enter Last name"
                          required={true}
                          type="text"
                        />
                      </div>
                      <label>
                        Email Address
                      </label>
                      <div
                        className="col-xs-12 col-sm-12"
                      >
                        <input
                          className="input-block"
                          id="email"
                          name="email"
                          onChange={[Function]}
                          placeholder="Enter email address"
                          required={true}
                          type="email"
                        />
                      </div>
                      <label>
                        Password
                      </label>
                      <div
                        className="col-xs-12 col-sm-12 col-lg-12"
                      >
                        <input
                          className="input-block"
                          id="password"
                          name="password"
                          onChange={[Function]}
                          placeholder="Enter password"
                          required={true}
                          type="password"
                        />
                      </div>
                      <label>
                        Confirm Password
                      </label>
                      <div
                        className="col-sx-12 col-sm-12 col-lg-12"
                      >
                        <input
                          className="input-block"
                          id="cpassword"
                          name="cpassword"
                          onChange={[Function]}
                          placeholder="Enter password"
                          required={true}
                          type="password"
                        />
                      </div>
                      <div
                        className="col-xs-12 col-sm-12 col-lg-12"
                      >
                        <button
                          className="btn btn-pink btn-full"
                          type="submit"
                        >
                          Sign Up
                        </button>
                        <br />
                        <br />
                        <Link
                          className="sign"
                          to="/signin"
                        >
                          <a
                            className="sign"
                            href="/signin"
                            onClick={[Function]}
                          >
                            Don't have an account? 
                            <span
                              className="btn-yellow-test"
                            >
                              [SIGN IN]
                            </span>
                          </a>
                        </Link>
                      </div>
                    </div>
                  </div>
                </form>
              </div>
            </div>
          </SignUpModal>
        </Connect(SignUpModal)>
      </withRouter(Connect(SignUpModal))>
    </Router>
  </BrowserRouter>
</Provider>
`;
